### ISSUE TITLE
Symbols-view integration is missing

### ISSUE THREAD

**Issue by diminutivesloop** [TypeStrong/atom-typescript#1219]
Before the 11.0.1 update I could access Typescript symbols via the symbols-view package's toggle-symbols and toggle-project-symbols commands. This was a particularly convenient feature for navigating large codebases and large files, so I'd really appreciate if it can be fixed or re-implemented.

**Comment by guncha** [comment]
Yeah, it's one of the things that didn't make it over, sorry. I don't even know how it could be reimplemented since we don't have direct access to the Typescript AST anymore, but I'm sure it's possible. If there's enough demand for it, might be worth it.

**Comment by dominic31** [comment]
+1, would be great to get this back.

**Comment by robofish** [comment]
Would also be lovely to get back support for `symbols-view:go-to-declaration` / the "Go To Declaration" in the context menu 😀

**Comment by manuelnaturalcycles** [comment]
@guncha can you point me in the direction needed to implement this? i would like to give it a try because this seriously breaks atom for me. i cant really work without go to declaration keybind, i work in a massive typescript project.

thanks and sorry for bothering you.

**Comment by guncha** [comment]
@manuelnaturalcycles, we have a `Go to declaration` command that is bound to `F12`. I guess there's a `Symbols view: Go to declaration` that doesn't do anything. I should get rid of that. It's not in the context menu though, is that the problem?

If you mean implementing Symbols view, it's going to be hard, because AFAIK there's no way to access all the symbols for a given file via the `tsserver` API. You could ask for `quickInfo` for every location in the file, but that would be madness.

**Comment by manuelnaturalcycles** [comment]
@gunchaI i see.
i hate myself for not having the knowledge to approach this. 
i remember having this keybind in atom:
`'atom-workspace atom-text-editor:not([mini])':
  'cmd-g': 'symbols-view:go-to-declaration'`
and it worked. i wondered if it was possible to do some work without going too deep to bring that functionality back. but it seems to be pretty deep into things i dont even know of :(

**Comment by andregomars** [comment]
+1, need native symbol view support of .ts files

**Comment by mashleyttu** [comment]
Would love to have this feature back. Used it consistently. :)

**Comment by russaa** [comment]
maybe the ["navtree" command](https://github.com/Microsoft/TypeScript/blob/master/lib/protocol.d.ts#L1623) from the `tsserver` API could be used for creating a symbols-list/-tree.

A quick test with a small example:
```typescript
export class Greeter {
    greeting: string;
    constructor(message: string) {
        this.greeting = message;
    }
    greet() {
        return "Hello, " + this.greeting;
    }
}

// let greeter = new Greeter("world");
```

... and the `navtree` command & response:
```json
> tsserver
{"seq":1,"type":"quickinfo","command":"open","arguments":{"file":"C:/test_tsserver/src/greeter.ts"}}
{"seq":1,"type":"quickinfo","command":"navtree","arguments":{"file":"C:/test_tsserver/src/greeter.ts"}}

{
 "seq": 0,
 "type": "response",
 "command": "navtree",
 "request_seq": 1,
 "success": true,
 "body": {
  "text": "\"greeter\"",
  "kind": "module",
  "kindModifiers": "",
  "spans": [{
    "start": {
     "line": 1,
     "offset": 1
    },
    "end": {
     "line": 11,
     "offset": 39
    }
   }
  ],
  "childItems": [{
    "text": "Greeter",
    "kind": "class",
    "kindModifiers": "export",
    "spans": [{
      "start": {
       "line": 1,
       "offset": 1
      },
      "end": {
       "line": 9,
       "offset": 2
      }
     }
    ],
    "childItems": [{
      "text": "constructor",
      "kind": "constructor",
      "kindModifiers": "",
      "spans": [{
        "start": {
         "line": 3,
         "offset": 5
        },
        "end": {
         "line": 5,
         "offset": 6
        }
       }
      ]
     }, {
      "text": "greet",
      "kind": "method",
      "kindModifiers": "",
      "spans": [{
        "start": {
         "line": 6,
         "offset": 5
        },
        "end": {
         "line": 8,
         "offset": 6
        }
       }
      ]
     }, {
      "text": "greeting",
      "kind": "property",
      "kindModifiers": "",
      "spans": [{
        "start": {
         "line": 2,
         "offset": 5
        },
        "end": {
         "line": 2,
         "offset": 22
        }
       }
      ]
     }
    ]
   }
  ]
 }
}
```

**Comment by guncha** [comment]
Interesting, without knowing exact details a symbol view would have, that seems like it would fit the bill.

**Comment by vladeck** [comment]
Looking at the VSCode's source code (`workspaceSymbolProvider.ts`), I think the solution lays in [navto](https://github.com/Microsoft/vscode/blob/master/extensions/typescript/src/features/workspaceSymbolProvider.ts#L62) command for `tsserver`.

What is the starting point in atom-typescript source code where one would look in order to figure out how to implement functional `symbols-view:toggle-file-symbols`?

**Comment by guncha** [comment]
@vladeck, you'd have to find out how to integrate with the Atom's symbol view. It might be through a provider in which case look at the autocomplete provider as an example. If it's just a command, you can add a new command by looking at the examples in the `commands` directory.

**Comment by danyright** [comment]
I'm new to Atom and TypeScript, but I see that symbols(-tree) supports ActionScript 3 with package ["language-actionscript3"](https://atom.io/packages/language-actionscript3) which is very similar to TypeScript in it's structure (based on ECMAScript). Wouldn't it be possible to use this as an example to get it working with TS? Because it is indeed a very useful feature on larger files.

**Comment by quyin** [comment]
FYI: Because symbols-view uses ctags, a temporary workaround would be to add typescript support to ctags, e.g., using https://github.com/jmquigley/typescript-ctags

**Comment by guikubivan** [comment]
I got confused by the Symbols View "Go to Declaration" option in the context menu(which doesn't do anything), so I thought it was removed. Good to know it still works with F12. Only discovered that through this ticket unfortunately. Meanwhile, I've disabled Symbols View for my sanity.

**Comment by russaa** [comment]
I started an implementation for the file-symbols (i.e. only symbols from current file/active editor, not from the project).

the repo/branch is (it also includes the semantic-view, as that was the feature I was originally working on)
[https://github.com/mmig/atom-typescript/tree/issue1279_semantic-view_and_file-symbols](https://github.com/mmig/atom-typescript/tree/issue1279_semantic-view_and_file-symbols)

It is basically a copy/excerpt from the `symbols-view` (system package) but instead of using the ctag-parser, it uses the tsserver API extracting & using the navTree.

In principal, it works, but either would need different keybindings, or the keybindings for `symbols-view` need to be overwritten (for editors with typescript content), e.g. something like:
```
'atom-text-editor:not([mini]):not([data-grammar^="source ts"])':
  'ctrl-shift-O': 'symbols-view:toggle-file-symbols'

'atom-text-editor:not([mini])[data-grammar^="source ts"]':
  'ctrl-shift-O': 'typescript:toggle-file-symbols'
```

here is what it looks like (exactly the same as the `symbols-view`):
![image](https://user-images.githubusercontent.com/3603062/35759157-2b9fcc94-0878-11e8-9296-515f61874a34.png)

**Comment by lierdakil** [comment]
> keybindings for symbols-view need to be overwritten

FWIW, that's pretty much [what I did](
https://github.com/atom-haskell/ide-haskell-hasktags/blob/a130949eaa705e3090e7302190ca000babd71f5e/keymaps/ide-haskell-hasktags.cson) when making the thing for Haskell. You don't need the `:not` thing though since the more specific selector wins. And `atom-text-editor:not([mini])[data-grammar^="source ts"]` is more specific than `atom-text-editor:not([mini])`

Although the whole idea about binding to syntax scope is rather brittle, for example we have `ts` and `tsx` source scopes. I would personally prefer if we set a class on the text editors atom-typescript manages and matched on that.

**Comment by russaa** [comment]
yes, I agree:  
although `tsx` currently works with `[data-grammar^="source ts"]`, the selection should be more deliberate.

The typescript editor is currently marked with the class `typescript-editor`: would that be an appropriate selector?

**Comment by szhu** [comment]
Easy instructions in case anyone wants to try out [the experimental branch mentioned above](#issuecomment-362738456)!

```sh
cd ~/.atom/packages/atom-typescript
git init
git clean -fd  # This deletes any local changes you've made to the plugin!
git remote add mmig/atom-typescript https://github.com/mmig/atom-typescript.git
git fetch mmig/atom-typescript
git checkout mmig/atom-typescript/issue1279_semantic-view_and_file-symbols
```

**Comment by russaa** [comment]
I added search for project-symbols:  
it's done via `"navto"` as @vladeck suggested

but there are some caveats:  
 * in difference to the project-symbols via `symbol-view`, there needs to be at least one letter/char in the search-field, otherwise no results/empty list will be shown (the [doc claims](https://github.com/Microsoft/TypeScript/blob/9f10888b10df8e5d8a35f37f809fbc6f3a32c987/src/server/protocol.ts#L2252) that `.*` can be used, but I always get an empty list for that `searchValue`)
 * I could not overwrite the default keybindings for this, so that it would open when a TS-editor is focused:  
    `symbol-view` registers `ctrl-shift-r` directly on `.platform-win32, .platform-linux` and that seems to prevent/circumvent the "more-specific precedes more-general" rule
   (is there a way to disable that global keybinding in case a specific element/editor is currently focused?)

As a result, the project-symbols are not generated once then filtered (as in `symbol-view`), but instead on every change of the search-string, a new `"navto"` query is triggered.

B.t.w. is there a way to cancel/abort a pending requested that was started via the `clientResolver`/`client` implementation?  
It would probably be a good thing to cancel a pending `"navto"` query, before starting a new one...

**Comment by russaa** [comment]
@lierdakil  
this can probably be closed now -- seeing that the symbols-view alternative is included since version 12.3.0

**Comment by lierdakil** [comment]
Right.
